scio-avro/src/main/scala/com/spotify/scio/avro/AvroIO.scala

```diff
@@ -18,422 +18,316 @@
+    codec: CodecFactory = WriteParam.DefaultCodec,
+    metadata: Map[String, AnyRef] = WriteParam.DefaultMetadata,
+    filenamePolicySupplier: FilenamePolicySupplier = WriteParam.DefaultFilenamePolicySupplier,
+    prefix: String = WriteParam.DefaultPrefix,
+    shardNameTemplate: String = WriteParam.DefaultShardNameTemplate,
+    tempDirectory: String = WriteParam.DefaultTempDirectory
+  )
 }
 
-object ObjectFileIO {
+final case class GenericRecordIO(
+  path: String,
+  schema: Schema,
+  datumFactory: AvroDatumFactory[GenericRecord] = GenericRecordDatumFactory
+) extends AvroIO[GenericRecord] {
+  override protected def baseRead: BAvroIO.Read[GenericRecord] =
+    BAvroIO.readGenericRecords(schema)
+  override protected def baseWrite: BAvroIO.Write[GenericRecord] =
+    BAvroIO.writeGenericRecords(schema)
+  override def tap(read: ReadP): Tap[GenericRecord] =
+    GenericRecordTap(path, schema, datumFactory, read)
+}
+
+object GenericRecordIO {
   type ReadParam = AvroIO.ReadParam
   val ReadParam = AvroIO.ReadParam
   type WriteParam = AvroIO.WriteParam
   val WriteParam = AvroIO.WriteParam
 }
 
-final case class ProtobufIO[T <: Message: ClassTag](path: String) extends ScioIO[T] {
-  override type ReadP = ProtobufIO.ReadParam
-  override type WriteP = ProtobufIO.WriteParam
-  final override val tapT: TapT.Aux[T, T] = TapOf[T]
-  private val protoCoder = Coder.protoMessageCoder[T]
-
-  /**
-   * Get an SCollection for a Protobuf file.
-   *
-   * Protobuf messages are serialized into `Array[Byte]` and stored in Avro files to leverage Avro's
-   * block file format.
-   */
-  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
-    sc.read(ObjectFileIO[T](path)(protoCoder))(params)
-
-  /**
-   * Save this SCollection as a Protobuf file.
-   *
-   * Protobuf messages are serialized into `Array[Byte]` and stored in Avro files to leverage Avro's
-   * block file format.
-   */
-  override protected def write(data: SCollection[T], params: WriteP): Tap[T] = {
-    val metadata = params.metadata ++ ProtobufUtil.schemaMetadataOf[T]
-    data.write(ObjectFileIO[T](path)(protoCoder))(params.copy(metadata = metadata)).underlying
-  }
+final case class SpecificRecordIO[T <: SpecificRecord: ClassTag](
+  path: String,
+  datumFactory: AvroDatumFactory[T]
+) extends AvroIO[T] {
 
-  override def tap(read: ReadP): Tap[T] =
-    ObjectFileTap[T](path, read)(protoCoder)
+  private val recordClass: Class[T] = ScioUtil.classOf[T]
+  override def schema: Schema = SpecificData.get().getSchema(recordClass)
+  override def baseRead: BAvroIO.Read[T] = BAvroIO.read(recordClass)
+  override def baseWrite: BAvroIO.Write[T] = BAvroIO.write(recordClass)
+  override def tap(read: ReadP): Tap[T] = SpecificRecordTap(path, datumFactory, read)
 }
 
-object ProtobufIO {
+object SpecificRecordIO {
+  private[scio] def defaultDatumFactory[T <: SpecificRecord: ClassTag]: AvroDatumFactory[T] =
+    new SpecificRecordDatumFactory[T](ScioUtil.classOf[T])
```

Comment:
@shnapz using this factory, we'll load the model and conversion from the generated code + we'll force `CharSequence` to get `String` as implementation